ci/lint.dart

```diff
@@ -5,17 +5,279 @@
+    this.workingDirectory,
+    this.printOutput = false,
+  });
+
+  /// The name of the job.
+  final String name;
+
+  /// The arguments for the process, including the command name as args[0].
+  final List<String> args;
+
+  /// The working directory that the command should be executed in.
+  final Directory? workingDirectory;
+
+  /// Whether or not this command should print it's stdout when it runs.
+  final bool printOutput;
+
+  @override
+  String toString() {
+    return args.join(' ');
+  }
+}
+
+/// A pool of worker processes that will keep [numWorkers] busy until all of the
+/// (presumably single-threaded) processes are finished.
+class ProcessPool {
+  ProcessPool({int? numWorkers})
+      : numWorkers = numWorkers ?? Platform.numberOfProcessors;
+
+  ProcessRunner processRunner = ProcessRunner();
+  int numWorkers;
+  List<WorkerJob> pendingJobs = <WorkerJob>[];
+  List<WorkerJob> failedJobs = <WorkerJob>[];
+  Map<WorkerJob, Future<List<int>>> inProgressJobs =
+      <WorkerJob, Future<List<int>>>{};
+  Map<WorkerJob, ProcessRunnerResult> completedJobs =
+      <WorkerJob, ProcessRunnerResult>{};
+  Completer<Map<WorkerJob, ProcessRunnerResult>> completer =
+      Completer<Map<WorkerJob, ProcessRunnerResult>>();
+
+  void _printReport() {
+    final int totalJobs =
+        completedJobs.length + inProgressJobs.length + pendingJobs.length;
+    final String percent = totalJobs == 0
+        ? '100'
+        : ((100 * completedJobs.length) ~/ totalJobs).toString().padLeft(3);
+    final String completed = completedJobs.length.toString().padLeft(3);
+    final String total = totalJobs.toString().padRight(3);
+    final String inProgress = inProgressJobs.length.toString().padLeft(2);
+    final String pending = pendingJobs.length.toString().padLeft(3);
+    stdout.write(
+        'Jobs: $percent% done, $completed/$total completed, $inProgress in progress, $pending pending.  \r');
+  }
+
+  Future<List<int>> _scheduleJob(WorkerJob job) async {
+    final Completer<List<int>> jobDone = Completer<List<int>>();
+    final List<int> output = <int>[];
+    try {
+      completedJobs[job] = await processRunner.runProcess(
+        job.args,
+        workingDirectory: job.workingDirectory,
+        printOutput: job.printOutput,
+      );
+    } catch (e) {
+      failedJobs.add(job);
+      if (e is ProcessRunnerException) {
+        print(e.toString());
+        print('${utf8.decode(output)}');
+      } else {
+        print('\nJob $job failed: $e');
+      }
+    } finally {
+      inProgressJobs.remove(job);
+      if (pendingJobs.isNotEmpty) {
+        final WorkerJob newJob = pendingJobs.removeAt(0);
+        inProgressJobs[newJob] = _scheduleJob(newJob);
+      } else {
+        if (inProgressJobs.isEmpty) {
+          completer.complete(completedJobs);
+        }
+      }
+      jobDone.complete(output);
+      _printReport();
+    }
+    return jobDone.future;
+  }
+
+  Future<Map<WorkerJob, ProcessRunnerResult>> startWorkers(
```

Comment:
nit: This would be more efficient and robust if it returned `Stream<Tuple<WorkerJob, ProcessRunnerResult>>`.

Reply:
Cool, thanks for the suggestion, I'll do that.